### PR TITLE
TF-2101 [Part 4] Handle restore deleted message

### PR DESCRIPTION
**Pull request by hieutbui**
## Issue
---
#2101 
#2102 

### Need merge
---
https://github.com/linagora/tmail-flutter/pull/2411

## Resolved
---
- Web

https://github.com/linagora/tmail-flutter/assets/80142234/245fec4b-c790-424d-a47b-c7b7f5f965f7

- Android

https://github.com/linagora/tmail-flutter/assets/80142234/6fc7ee87-04b2-4c83-81c4-782477689b7b

### CODE REVIEW

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 1]
```diff
@@ -264,6 +265,8 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
           _refreshMailboxChanges(currentMailboxState: success.currentMailboxState);
         } else if (success is EmptySpamFolderSuccess) {
           _refreshMailboxChanges(currentMailboxState: success.currentMailboxState);
+        } else if (success is RestoreDeletedMessageSuccess) {
+          refreshAllMailbox();
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:269]: why not `_refreshMailboxChanges`?

File: lib/features/email_recovery/presentation/email_recovery_controller.dart [hunk 2]
```diff
@@ -285,6 +294,45 @@ class EmailRecoveryController extends BaseController with DateRangePickerMixin {
     }
   }
 
+  void onRestore(BuildContext context) {
+    KeyboardUtils.hideKeyboard(context);
+    
+    final UTCDate? deletedBefore;
+    final UTCDate? deletedAfter;
+    final UTCDate? receivedBefore;
+    final UTCDate? receivedAfter;
```
Comment (dab246) [inline lib/features/email_recovery/presentation/email_recovery_controller.dart:303]: Don't use final if you want to change the value of the variable.

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 3]
```diff
@@ -2225,28 +2251,93 @@ class MailboxDashBoardController extends ReloadableController {
     setSelectedEmail(email.toPresentationEmail());
   }
 
+  void _handleRestoreDeletedMessageSuccess(EmailRecoveryActionId? emailRecoveryActionId) {
+    log('MailboxDashBoardController::_handleRestoreDeletedMessageSuccess():emailRecoveryActionId: $emailRecoveryActionId');
+    _getAllMailboxInteractor = getBinding<GetAllMailboxInteractor>();
+    _getAllMailboxInteractor?.execute(sessionCurrent!, accountId.value!);
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2257]: why call get all mailbox at here?

File: lib/features/mailbox_dashboard/presentation/widgets/recover_deleted_message_loading_banner_widget.dart [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      (success) {
+        if (success is RestoreDeletedMessageLoading 
+          || success is RestoreDeletedMessageSuccess 
+          || success is GetRestoredDeletedMessageLoading
+        ) {
+          return Padding(
+            padding: EdgeInsets.only(
+              top: responsiveUtils.isDesktop(context) ? 16 : 0,
+              right: 16,
+              bottom: responsiveUtils.isDesktop(context) ? 0 : 16,
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/widgets/recover_deleted_message_loading_banner_widget.dart:41]: Please use EdgeInsetsDirectional

**Comment by dab246** [comment]
### Note 

Screenshot 2024-01-02 at 15 00 38

File: lib/features/email_recovery/presentation/email_recovery_controller.dart [hunk 5]
```diff
@@ -1,29 +1,145 @@
...
+  final senderExpandMode = ExpandMode.EXPAND.obs;
+
   final focusManager = InputFieldFocusManager.initial();
 
   List<EmailAddress> listRecipients = <EmailAddress>[];
   List<EmailAddress> listSenders = <EmailAddress>[];
 
+  TextEditingController subjectFieldInputController = TextEditingController();
+  TextEditingController recipientsFieldInputController = TextEditingController();
+  TextEditingController senderFieldInputController = TextEditingController();
```
Comment (dab246) [inline lib/features/email_recovery/presentation/email_recovery_controller.dart:53]: Please dispose it

File: lib/features/email_recovery/presentation/widgets/text_input_field/avatar_suggestion_item_widget.dart [hunk 6]
```diff
@@ -0,0 +1,35 @@
+import 'package:core/core.dart';
```
Comment (dab246) [inline lib/features/email_recovery/presentation/widgets/text_input_field/avatar_suggestion_item_widget.dart:1]: Only import class necessary

**Comment by hieutbui** [comment]
> ### Note
> Screenshot 2024-01-02 at 15 00 38


https://github.com/linagora/tmail-flutter/assets/80142234/4321a6b4-ce20-4e86-a6cd-efc47bbe46eb


Please check this video!

File: lib/features/email_recovery/presentation/email_recovery_controller.dart (hunk 5 above)
Comment (hieutbui) [reply lib/features/email_recovery/presentation/email_recovery_controller.dart]: done

File: lib/features/email_recovery/presentation/widgets/text_input_field/avatar_suggestion_item_widget.dart (hunk 6 above)
Comment (hieutbui) [reply lib/features/email_recovery/presentation/widgets/text_input_field/avatar_suggestion_item_widget.dart]: done

**Comment by dab246** [comment]
> > ### Note
> > Screenshot 2024-01-02 at 15 00 38
> 
>  after.recovered.mp4 
> Please check this video!

Please check the Desc of the issue. What actions does Folder Recovered only have? See below

Screenshot 2024-01-04 at 18 08 01

Screenshot 2024-01-04 at 18 08 12

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 7]
```diff
@@ -499,4 +502,51 @@ class MailboxAPI with HandleSetErrorMixin {
...
+    final getMailboxInvocation = requestBuilder.invocation(getMailBoxMethod);
+
+    final capabilities = getMailBoxMethod.requiredCapabilities
+      .toCapabilitiesSupportTeamMailboxes(session, accountId);
+
+    final result = await (requestBuilder
+            ..usings(capabilities))
+          .build()
+          .execute();
+
```
Comment (dab246) [inline lib/features/mailbox/data/network/mailbox_api.dart:537]: indent

File: lib/features/mailbox/data/network/mailbox_api.dart [hunk 8]
```diff
@@ -499,4 +502,51 @@ class MailboxAPI with HandleSetErrorMixin {
...
+    
+    final mailboxResponse = result.parse<GetMailboxResponse>(
+      getMailboxInvocation.methodCallId,
+      GetMailboxResponse.deserialize
+    );
+
+    if (mailboxResponse?.list.isNotEmpty == true) {
+      return GetMailboxByRoleResponse(
+        mailbox: mailboxResponse!.list.first,
+        mailboxState: mailboxResponse.state
```
Comment (dab246) [inline lib/features/mailbox/data/network/mailbox_api.dart:546]: IMO, we don't need it. If you want refresh changes mailbox , you can use the mailbox's `state_cache`. Use `getMailboxState` in `MailboxRepository`

**Comment by hieutbui** [comment]
> > > ### Note
> > > Screenshot 2024-01-02 at 15 00 38
> > 
> > 
> > after.recovered.mp4
> > Please check this video!
> 
> Please check the Desc of the issue. What actions does Folder Recovered only have? See below
> 
> Screenshot 2024-01-04 at 18 08 01 Screenshot 2024-01-04 at 18 08 12

I fixed it!
image

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 9]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
+  void _handleRestoreDeletedMessageSuccess(EmailRecoveryActionId? emailRecoveryActionId) async {
+    log('MailboxDashBoardController::_handleRestoreDeletedMessageSuccess():emailRecoveryActionId: $emailRecoveryActionId');
+    consumeState(
+      _getMailboxByRoleInteractor.execute(
+        sessionCurrent!,
+        accountId.value!,
+        Role('Restored-Messages')
+      )
+    );
+    await Future.delayed(const Duration(seconds: 1));
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2268]: why we need delayed time at here?

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 10]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
+    if (emailRecoveryActionId != null) {
+      consumeState(_getRestoredDeletedMessageInteractor.execute(emailRecoveryActionId));
```
Comment (dab246) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2270]: What is `GetRestoredDeletedMessageInteractor ` ?

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 11]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
 
+  void _handleRestoreDeletedMessageSuccess(EmailRecoveryActionId? emailRecoveryActionId) async {
+    log('MailboxDashBoardController::_handleRestoreDeletedMessageSuccess():emailRecoveryActionId: $emailRecoveryActionId');
+    consumeState(
+      _getMailboxByRoleInteractor.execute(
+        sessionCurrent!,
+        accountId.value!,
+        Role('Restored-Messages')
+      )
+    );
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2267]: I still not understand the concept of this interactor, please move it out of this logic

Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2267]: dont combine `consumeState` in controller

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 12]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
+        AppLocalizations.of(currentOverlayContext!).restoreDeletedMessageFailed
+      );
+  }
+
+  void _handleGetRestoredDeletedMessageSuccess(EmailRecoveryAction emailRecoveryAction) async {
+    switch (emailRecoveryAction.status) {
+      case EmailRecoveryStatus.inProgress:
+      case EmailRecoveryStatus.waiting:
+        await Future.delayed(const Duration(seconds: 1));
+        _handleRestoreDeletedMessageSuccess(emailRecoveryAction.id);
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2288]: not make sense

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 13]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
+            leadingSVGIconColor: Colors.white,
+            backgroundColor: AppColor.toastSuccessBackgroundColor,
+            textColor: Colors.white,
+            actionIcon: SvgPicture.asset(
+              imagePaths.icFolderMailbox,
+              colorFilter: Colors.white.asFilter(),
+            )
+          );
+        } else {
+          _handleRestoreDeletedMessageFailed();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2309]: not make sense

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 14]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
+        }
+        break;
+      default:
+        _handleRestoreDeletedMessageFailed();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2313]: not make sense

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart (hunk 11 above)
Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2267]: move the logic to get `recovered` folder to the get status of recovery task (only when complete)

File: lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart [hunk 15]
```diff
@@ -2225,28 +2256,95 @@ class MailboxDashBoardController extends ReloadableController {
...
+        currentOverlayContext!,
+        AppLocalizations.of(currentOverlayContext!).restoreDeletedMessageFailed
+      );
+  }
+
+  void _handleGetRestoredDeletedMessageSuccess(EmailRecoveryAction emailRecoveryAction) async {
+    switch (emailRecoveryAction.status) {
+      case EmailRecoveryStatus.inProgress:
+      case EmailRecoveryStatus.waiting:
+        await Future.delayed(const Duration(seconds: 1));
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/controller/mailbox_dashboard_controller.dart:2287]: delay 2 second before call the status again

File: lib/features/email/domain/usecases/get_restored_deleted_message_interactor.dart [hunk 16]
```diff
@@ -4,21 +4,52 @@ import 'package:core/presentation/state/failure.dart';
...
+
+          yield Right<Failure, Success>(GetRestoredDeletedMessageSuccess(emailRecovery, recoveredMailbox: recoveredMailbox));
+          break;
+        case EmailRecoveryStatus.failed:
+          yield Left<Failure, Success>(GetRestoredDeletedMessageFailure(null));
+          break;
+        case EmailRecoveryStatus.inProgress:
+        case EmailRecoveryStatus.waiting:
+        case EmailRecoveryStatus.canceled:
+          yield Right<Failure, Success>(GetRestoredDeletedMessageSuccess(emailRecovery));
```
Comment (hoangdat) [inline lib/features/email/domain/usecases/get_restored_deleted_message_interactor.dart:48]: Why not separate like this:
- `GetRestoredDeletedMessageInProgress`
- `GetRestoredDeletedMessageWaiting`
- `GetRestoredDeletedMessageCancelled`

File: lib/l10n/intl_messages.arb [hunk 17]
```diff
@@ -3594,10 +3594,16 @@
     "placeholders_order": [],
     "placeholders": {}
   },
-  "bannerProgressingRecoveryMessage": "The recovery is in progress. You can continue using TeamMail",
+  "bannerProgressingRecoveryMessage": "The recovery is in progress. You can continue using TwakeMail",
```
Comment (hoangdat) [inline lib/l10n/intl_messages.arb:3597]: Twake Mail

File: lib/features/mailbox/presentation/mailbox_controller.dart [hunk 18]
```diff
@@ -283,7 +283,7 @@ class MailboxController extends BaseMailboxController with MailboxActionHandlerM
         mailboxDashBoardController.clearMailboxUIAction();
       } else if (action is RefreshChangeMailboxAction) {
         if (action.newState != currentMailboxState) {
-          _refreshMailboxChanges();
+          _refreshMailboxChanges(properties: action.properties);
```
Comment (hoangdat) [inline lib/features/mailbox/presentation/mailbox_controller.dart:286]: good catch, maybe resolve many issues :))

**Review by hoangdat — APPROVED** [review]

File: lib/features/mailbox/presentation/mailbox_controller.dart (hunk 18 above)
Comment (dab246) [reply lib/features/mailbox/presentation/mailbox_controller.dart:286]: I don't think we need this.

File: lib/features/mailbox/presentation/action/mailbox_ui_action.dart [hunk 19]
```diff
@@ -16,11 +17,12 @@ class SelectMailboxDefaultAction extends MailboxUIAction {}
 
 class RefreshChangeMailboxAction extends MailboxUIAction {
   final jmap.State? newState;
+  final Properties? properties;
```
Comment (dab246) [inline lib/features/mailbox/presentation/action/mailbox_ui_action.dart:20]: Why we need it?

**Review by dab246 — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**